moto/ec2/models/fleets.py

```diff
@@ -0,0 +1,311 @@
+                security_group_names=launch_spec.group_set,
+                user_data=launch_spec.user_data,
+                placement=None,
+                kernel_id=None,
+                ramdisk_id=None,
+                monitoring_enabled=launch_spec.monitoring,
+                subnet_id=launch_spec.subnet_id,
+                fleet_id=self.id,
+                tags=launch_spec.tag_specifications,
+            )
+
+            # get the instance from the reservation
+            instance = reservation.instances[0]
+            self.on_demand_instances.append(
+                {
+                    "id": reservation.id,
+                    "instance": instance,
+                }
+            )
+        self.fulfilled_capacity += added_weight
+        return self.on_demand_instances
+
+    def get_launch_spec_counts(self, weight_to_add):
+        weight_map = defaultdict(int)
+
+        weight_so_far = 0
+        if (
+            self.spot_options
+            and self.spot_options["AllocationStrategy"] == "diversified"
+        ):
+            launch_spec_index = 0
+            while True:
+                launch_spec = self.launch_specs[
+                    launch_spec_index % len(self.launch_specs)
+                ]
+                weight_map[launch_spec] += 1
+                weight_so_far += launch_spec.weighted_capacity
+                if weight_so_far >= weight_to_add:
+                    break
+                launch_spec_index += 1
+        else:  # lowestPrice
+            cheapest_spec = sorted(
+                # FIXME: change `+inf` to the on demand price scaled to weighted capacity when it's not present
+                self.launch_specs,
+                key=lambda spec: float(spec.spot_price or "+inf"),
+            )[0]
+            weight_so_far = weight_to_add + (
+                weight_to_add % cheapest_spec.weighted_capacity
+            )
+            weight_map[cheapest_spec] = int(
+                weight_so_far // cheapest_spec.weighted_capacity
+            )
+
+        return weight_map, weight_so_far
+
+    def terminate_instances(self):
+        instance_ids = []
+        new_fulfilled_capacity = self.fulfilled_capacity
+        for req in self.spot_requests + self.on_demand_instances:
+            instance = None
+            try:
+                instance = req.instance
+            except AttributeError:
+                instance = req["instance"]
+
+            if instance.state == "terminated":
+                continue
+
+            # stop when we hit the target capacity
+            if new_fulfilled_capacity <= self.target_capacity:
+                break
+
+            instance_ids.append(instance.id)
+            new_fulfilled_capacity -= 1
+
+        self.spot_requests = [
+            req for req in self.spot_requests if req.instance.id not in instance_ids
+        ]
+        self.on_demand_instances = [
+            req for req in self.on_demand_instances if req["id"] not in instance_ids
```

Comment:
What is the reason for using `req["id"]` here, which refers to the reservation ID? Why not `req["instance"].id`?